### PR TITLE
Fix Caret Height On Empty Lines

### LINKED ISSUE #92080: The cursor height becomes shorter when I delete all characters from TextField
like the following picture:

<details>
<summary>flutter doctor -v</summary>

```log
[√] Flutter (Channel stable, 2.5.3, on Microsoft Windows [Version 10.0.17763.2237], locale    en-US)
    • Flutter version 2.5.3 at D:\SDK\Flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • FLUTTER_GIT_URL = https://gitee.com/mirrors/Flutter.git
    • Framework revision 18116933e7 (4 days ago), 2021-10-15 10:46:35 -0700
    • Engine revision d3ea636dc5
    • Dart version 2.14.4
    • Pub download mirror https://pub.flutter-io.cn
    • Flutter download mirror https://storage.flutter-io.cn

[√] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at D:\SDK\Android
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = D:\SDK\Android
    • ANDROID_SDK_ROOT = D:\SDK\Android
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6842174)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • CHROME_EXECUTABLE = C:\Program Files (x86)\Microsoft\Edge\Application\msedge.exe

[√] Visual Studio - develop for Windows (Visual Studio Community 2019 16.11.5)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
    • Visual Studio Community 2019 version 16.11.31729.503
    • Windows 10 SDK version 10.0.19041.0

[√] Android Studio (version 4.2)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6842174)

[√] IntelliJ IDEA Community Edition (version 2021.2)
    • IntelliJ at C:\Program Files\JetBrains\IntelliJ IDEA Community Edition 2021.2.3
    • Flutter plugin version 61.2.4
    • Dart plugin version 212.5486

[√] VS Code (version 1.55.2)
    • VS Code at C:\Users\Zhanghuifeng\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.15.0

[√] Connected device (3 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [Version 10.0.17763.2237]
    • Chrome (web)      • chrome  • web-javascript • unknown
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 94.0.992.50

• No issues found!
```
</details>

<details>
<summary>simple code</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'TextField Demo',
      theme: ThemeData(primarySwatch: Colors.teal),
      home: Scaffold(
        body: Center(
          child: SizedBox(
            width: 200,
            height: 80,
            child: TextField(
              controller: TextEditingController(text: "Hello"),
              minLines: 1,
              textAlignVertical: TextAlignVertical.center,
              cursorHeight: 14,
              decoration: InputDecoration(
                border: const OutlineInputBorder(),
                hintText: "HintText",
                labelText: "LabelText",
                hintStyle: const TextStyle(fontSize: 14),
                labelStyle: const TextStyle(fontSize: 15),
                counterText:  "",
              ),
            ),
          ),
        ),
      ),
    );
  }
}
```
</details>

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
Fixes https://github.com/flutter/flutter/issues/92080 (related: https://github.com/flutter/flutter/issues/114950)

- `getLocalRectForCaret` now reports the real rect that will be painted.
- Move caret x-coordinate clamping to RenderEditable since TextPainter doesn't know about clipping.
- Caret Caching no longer depends on "caret prototype".
- Fix `WidgetSpan.codeUnitAtVisitor` implementation

Incorporated some changes from https://github.com/flutter/flutter/pull/113268 and https://github.com/flutter/flutter/pull/118128

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/rendering/editable.dart [hunk 1]
```diff
@@ -1790,11 +1790,46 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin,
...
     _computeTextMetricsIfNeeded();
-    final Offset caretOffset = _textPainter.getOffsetForCaret(caretPosition, _caretPrototype);
-    // This rect is the same as _caretPrototype but without the vertical padding.
-    final Rect rect = Rect.fromLTWH(0.0, 0.0, cursorWidth, cursorHeight).shift(caretOffset + _paintOffset + cursorOffset);
-    // Add additional cursor offset (generally only if on iOS).
-    return rect.shift(_snapToPhysicalPixel(rect.topLeft));
+    final Rect caretPrototype = _caretPrototype;
+    final Offset caretOffset = _textPainter.getOffsetForCaret(caretPosition, caretPrototype);
+    Rect caretRect = caretPrototype.shift(caretOffset + cursorOffset);
+    final double scrollableWidth = math.max(_textPainter.width + _caretMargin, size.width);
```
Comment (Renzo-Olivares) [inline packages/flutter/lib/src/rendering/editable.dart:1796]: Not sure I understand why we are taking into account `scrollableWidth`, could explain for my sanity?

File: packages/flutter/lib/src/painting/text_painter.dart [hunk 2]
```diff
@@ -1166,8 +1196,8 @@ class TextPainter {
   ///
   /// Valid only after [layout] has been called.
   double? getFullHeightForCaret(TextPosition position, Rect caretPrototype) {
```
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/painting/text_painter.dart:1198]: It's no longer used. But as you pointed out it's a public API so I didn't change the signature.

File: packages/flutter/lib/src/painting/text_painter.dart [hunk 3]
```diff
@@ -273,14 +273,32 @@ class _UntilTextBoundary extends TextBoundary {
 /// This is used to cache and pass the computed metrics regarding the
 /// caret's size and position. This is preferred due to the expensive
 /// nature of the calculation.
-class _CaretMetrics {
-  const _CaretMetrics({required this.offset, this.fullHeight});
+///
+// This should be a sealed class: A _CaretMetrics is either a _LineCaretMetrics
```
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/painting/text_painter.dart:277]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#use-for-public-quality-private-documentation

The `//` part is not meant for API users just a maintenance note.

File: packages/flutter/lib/src/rendering/editable.dart (hunk 1 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/rendering/editable.dart:1796]: when there're trailing spaces the caret x offset may be greater than the width of the text painter.

**Review by Renzo-Olivares — APPROVED** [review]
LGTM, the broken tests seem unrelated probably just needs a sync up.

File: packages/flutter/lib/src/painting/text_painter.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (justinmc) [reply packages/flutter/lib/src/painting/text_painter.dart:1198]: Should we deprecate the parameter?

**Review by justinmc — APPROVED** [review]
LGTM 👍 

One comment about maybe adding a deprecation.

File: packages/flutter/lib/src/painting/text_painter.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/painting/text_painter.dart:1198]: Yeah `FullHeight` and `Offset` are pretty limited. Maybe we'll have different caret painter classes and let them report the bounding box.

**Comment by LongCatIsLooong** [comment]
The google testing failures are expected. Merging manually.

**Pull request merged**